### DISCUSSION TITLE
Serializer on foreign key: Maximum call stack size exceeded ?

### DISCUSSION THREAD

**Discussion by dungkaka** [mikro-orm/mikro-orm#4768] [General]
I have two entities City and School, which have One-To-Many relationship. Describe beblow:

``` Typescript
@Entity()
export class City {
    @PrimaryKey()
    id: number;

    @Property({ unique: true })
    slug: string;

    @Property()
    name: string;

    @OneToMany(() => School, (b) => b.city)
    schools = new Collection<School>(this);

}

@Entity()
export class School {
    @PrimaryKey()
    id: number;

    @Property()
    name: string;

    @ManyToOne({
        entity: () => City,
        serializer: (value: Ref<City>) => {
            if (value.isInitialized()) return value;
            else return undefined;
        },
        ref: true,
        nullable: true
    })
    city: Ref<City>;

    @Property({ persist: false })
    cityId: number;

}

``` 

Now I want to find a city with populate schools, then I run code:
```Typescript
const city = await this.em.find(City, { schools: { id: 1 } },  { populate: ["schools"]  });
```

Then it throw error, the point is serializer on populate foreign key, what I want is return to client field "city" only populate it, otherwise return only "cityId" !
```Typescript
@ManyToOne({
      entity: () => City,
      serializer: (value: Ref<City>) => {
          if (value.isInitialized()) return value; // Raise error because of this,
          else return undefined;
      },
      ref: true,
      nullable: true
  })
  city: Ref<City>;
```

LOG_DEV_EXCEPTION RangeError: Maximum call stack size exceeded
    at Function.toObject (/Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/serialization/EntityTransformer.js:15:20)
    at School.prototype.toJSON (/Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/entity/EntityHelper.js:33:62)
    at WrappedEntity.toJSON (/Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/entity/WrappedEntity.js:47:28)
    at /Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/entity/ArrayCollection.js:35:67
    at Array.map (<anonymous>)
    at Collection.toArray (/Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/entity/ArrayCollection.js:35:32)
    at Function.processCollection (/Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/serialization/EntityTransformer.js:133:24)
    at Function.processProperty (/Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/serialization/EntityTransformer.js:94:38)
    at /Users/dungkaka/MyData/Work/Mapstudy/backend/main-backend/node_modules/@mikro-orm/core/serialization/EntityTransformer.js:50:43
    at Array.map (<anonymous>)
    

MikroORM version: 5.8.1
Enviroment: MacOS
Node: 16

**Comment by B4nan** [comment]
Returning an entity instance from inside serializer seems to be wrong as that will surely contain cycles, try this instead:

```ts
@ManyToOne({
    entity: () => City,
    serializer: (value: Ref<City>) => {
        if (value.isInitialized()) return wrap(value).toObject();
        else return undefined;
    },
    ref: true,
    nullable: true
})
city: Ref<City>;
```

**Comment by dungkaka** [comment]
Yeah, it works, it prevents loop, actually, I also want to hide the field "city" on array of populate schools, is there any way to do that ?
```JSON
{
    "id": 1,
    "name": "An Giang",
    "slug": "an-giang",
    "schools": [
        {
            "id": 1,
            "name": "Nguyễn  Công Trứ",
            "city": {          // hide it
                "id": 1,
                "name": "An Giang",
                "slug": "an-giang",
            },
            "cityId": 1
        },
        {
            "id": 2,
            "name": "Sở GD-ĐT",
            "city": {         // hide it
                "id": 1,
                "name": "An Giang",
                "slug": "an-giang",
            },
            "cityId": 1
        }
    ]
}

```

**Reply by B4nan** [reply]
I am not following, your serializer is returning value and you want to hide it instead?

**Comment by dungkaka** [comment]
I have two query: 

Query 1:
```Typescript
const city = await this.em.find(City, { schools: { id: 1 } },  { populate: ["schools"]  });
```
and query 2:
```Typescript
const school = await this.em.find(School, { cityId: 12 });  // it will not return "city" because of not populate as I want, it's okay !
```

When I add serializer to decorator, I mean query 2 only return "city" when it populated by { populate: ["city" }, it done as I want.
But I also run query 1 for some reason, now it will cause some unwanted behavior as discussion above. I wonder if there is any way can work satisfy both of them ?

**Reply by B4nan** [reply]
First of all, this approach of mapping one column to two properties is generally not supported. I keep updating things in that direction, but it's not a supported (nor documented) use case. And it's definitely not needed to do things this way, you should have only one property (the relation). It will be serialized to the PK if it's not populated, which is exactly what you are asking for.
